feat: Add a bunch of mips compilers

maringuu opened the pull request:
> Disclaimer: This is my first time using crosstool-NG

I'm not sure if this is the right way but from what I understand adding the configurations is the first thing to do for supporting new compilers.
If this is the right way I will add the configurations to the compiler-explorer repo too.
Currently, I only tested some compilers by using the `local_build.sh` scripts. 
Is there a way to build them with your CI?
I'm happy if you can point me to the right direction to get these compilers supported in, thanks!

Anyways here is the commit message:

Namely, all compilers that are supported by crosstool-ng but are missing here. I generated the configurations like so:

```sh
# First get the list samples
$ ct-ng list-samples
# Chose something that is approximatly what you want
$ ct-ng mips64-unknown-linux-gnu
# Now configure the right toolchain version
$ ct-ng menuconfig
```

dkm commented:
Hello!
You should be able to build them locally, for example by using the `local_build.sh` script. Having the config is nice, but that's not enough for a successful build :) It's often the case that it simply doesn't build.

Also, config should disable gdb (some samples enable that... that's wasted time/space) and the progress bar (screws up the log file).

dkm commented on build/latest/mips-10.5.0.config:
```diff
@@ -0,0 +1,963 @@
+# Additional supported languages:
+#
+CT_CC_LANG_CXX=y
+# CT_CC_LANG_FORTRAN is not set
```
You seem to only have C++ here. I tend to enable as many languages as possible.

dkm commented:
In order to update the config files, you can use something like that (pulled from my shell history, you'll need to adjust):
```bash
$ ./check_and_update_conf.py -a arm-unknown  --inplace --config-dir /path/to/compiler-explorer/compiler-explorer/etc/config --summary arm-unknown.sum --config-todo arm-unknown.todo --guess-previous --version 13.2.0 --create-api-tests arm-unknown-13.2.0.sh
```

This will read the corresponding ct-ng config, probe for known languages, find corresponding config file, checks if it can guess where to inject the new config. If it can find it, the config snippet will be written in the `todo` file. The `summary` file gives a summary (!) of what the script has been doing. And finally, the `test` script is a crude way of checking the compiler work... It uses the API to compile sample code for every lang + option (compile to asm, to object code, link to exec).

maringuu commented:
Thank you for your review, I addressed your comments like described in the following:

To simplify the review I decided to only add `gcc-9.5.0/mips-unknown-gnu-linux`.
I'll add the other ones afterward.

> You seem to only have C++ here. I tend to enable as many languages as possible.

In menuconfig I now enabled all "Additional supported languages" under "C compiler".
This would be C++ and Fortran.

> Also, config should disable gdb

Good catch. I now disabled everything under "Debug facilities". That would be duma, gdb, ltrace, and strace. Is there a reason to keep any of them?


Is there any guidance on when to enable `CT_STATIC_TOOLCHAIN`?
It seems that most `{arm,arm64}` toolchains have it enabled but I don't quite see the reason why..

> You should be able to build them locally, for example by using the local_build.sh script. 

I did but my question was more in the direction on how to build it on your hardware.
In the [documentation](https://github.com/compiler-explorer/infra/blob/main/BuildingCompilers.md) in compiler-explorer/infra it sounds like this should be possible.
But of course I'll build it locally first.

> update the config files

Thank you for the snipped. Worked like a charm, as you already saw in https://github.com/compiler-explorer/compiler-explorer/pull/5723 .

dkm commented:
> Thank you for your review, I addressed your comments like described in the following:
> 
> To simplify the review I decided to only add `gcc-9.5.0/mips-unknown-gnu-linux`. I'll add the other ones afterward.
> 
> > You seem to only have C++ here. I tend to enable as many languages as possible.
> 
> In menuconfig I now enabled all "Additional supported languages" under "C compiler". This would be C++ and Fortran.

Ok. I guess we could enable more, like objc, Ada, fortran... But no reason to spent too much time on getting them to work now.

> > Also, config should disable gdb
> 
> Good catch. I now disabled everything under "Debug facilities". That would be duma, gdb, ltrace, and strace. Is there a reason to keep any of them?

no, we don't provide any debug features (yet?).

> Is there any guidance on when to enable `CT_STATIC_TOOLCHAIN`? It seems that most `{arm,arm64}` toolchains have it enabled but I don't quite see the reason why..
> 
> > You should be able to build them locally, for example by using the local_build.sh script.
> 
> I did but my question was more in the direction on how to build it on your hardware. In the [documentation](https://github.com/compiler-explorer/infra/blob/main/BuildingCompilers.md) in compiler-explorer/infra it sounds like this should be possible. But of course I'll build it locally first.

You can't trigger builds in our infra sorry. The workflow is: modify the docker image (your current change) and when it's merged, an admin can trigger a build using this docker image.

> > update the config files
> 
> Thank you for the snipped. Worked like a charm, as you already saw in [compiler-explorer/compiler-explorer#5723](https://github.com/compiler-explorer/compiler-explorer/pull/5723) .

Nice! Thanks for your on getting more compilers installed :)

maringuu commented:
I added some architectures for gcc 9.5.0 as it seemed fitting for this pr.

How do you decide which ABI to enable for mips64?
In master mips64el always choses 32 and mips64 always 64. Is there a reason for this?

maringuu commented:
> I guess we could enable more, like objc, Ada, fortran

Yeah, I wondered why crosstool-ng did not let me chose any of these despite fortran.

> You can't trigger builds in our infra sorry. 
I expected that, but wondered how the PR should be tested. Anyways, I built all of them locally using ct-ng, but the build using local_build is still running.

maringuu commented:
I added some more compilers. If that's too much to review, feel free to ignore some commits and only cherrypick what you reviewed.

dkm commented:
> I added some architectures for gcc 9.5.0 as it seemed fitting for this pr.
> 
> How do you decide which ABI to enable for mips64? In master mips64el always choses 32 and mips64 always 64. Is there a reason for this?

We don't really decide anything. I guess we've based the work on the sample config files in ct-ng OR (haven't checked) it was discussed with the person who asked for the specific compiler.


> > I guess we could enable more, like objc, Ada, fortran
> 
> Yeah, I wondered why crosstool-ng did not let me chose any of these despite fortran.

Depends on the target. For -linux, you get more lang options than for a baremetal target (usually the runtimes depend on having a OS).

> > You can't trigger builds in our infra sorry.
> 
> I expected that, but wondered how the PR should be tested. Anyways, I built all of them locally using ct-ng, but the build using local_build is still running.

Ok. I'll test the build locally before merging, and then trigger the build after merging.

> 
> > Is there any guidance on when to enable CT_STATIC_TOOLCHAIN? It seems that most {arm,arm64} toolchains have it enabled but I don't quite see the reason why..
> 
> Any comment on this?

No, sorry. I think it's legacy setting... We don't review all setting of ct-ng, so it's possible some tuning can be made.

dkm commented:
I'll test the updated docker image (ubuntu 22.04) with your changes and see if that builds OK :)

dkm commented:
```
mips64el FAILED
mipsel FAILED
mips FAILED
mips64 FAILED
mips64el FAILED
mipsel FAILED
mips FAILED
mips64 FAILED
mips64el FAILED
mipsel FAILED
mips64 FAILED
mips FAILED
```
I would say that's not a success, but it's caused by an issue with the host compiler.

dkm commented:
```
252:mips64el SUCCESS
505:mipsel SUCCESS
758:mips SUCCESS
1011:mips64 SUCCESS
1254:mips64el SUCCESS
1497:mipsel SUCCESS
1740:mips SUCCESS
1983:mips64 SUCCESS
2226:mips64el SUCCESS
2469:mipsel SUCCESS
2712:mips64 SUCCESS
2955:mips SUCCESS
```
Looks good :D

dkm commented:
erm. Not really sure what happened here. I wanted to add an extra commit that disables the progress bar. Why is github closing the PR?

dkm commented:
Ok, driver error. I push `main` in your branch, so github considers the PR empty (because I made it so) so closes it. And of course now, I can't push proper revision in your branch, because the PR is closed. I'll push the commit directly

maringuu commented:
Thank you for your extra work and merging!
I'll update my pr over at the compiler explorer repo next week hopefully.

dkm commented:
Thanks for doing that. No rush, I still need to trigger the compiler build before we can merge the config anyway
